File: app/news.py
"""Headlines for each company, fetched on a schedule and shipped with the page.

The terminal reads the news service directly from the visitor's browser, which
is the honest way round: they see what it says, not what we say it said. But
that service limits how often one address may ask, and a visitor who opens
several companies quickly gets turned away.

So the keeper fetches the same headlines once an hour, slowly enough never to
be refused, and the page ships the result. A visitor sees headlines at once and
the browser refreshes them live on top. If the live call is refused, the shipped
ones stay, with their age shown.

Usage:
    python3 news.py
"""
import json
import os
import re
import time
import urllib.error
import urllib.parse
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
GDELT = "https://api.gdeltproject.org/api/v2/doc/doc"
GAP = 8            # seconds between calls, comfortably inside their limit
TRIES = 4
UA = {"User-Agent": "Mozilla/5.0", "Accept": "application/json"}

# The same queries the browser uses, so both sides ask the same question.
QUERY = {
    "ANTHROPIC": '"Anthropic" (IPO OR valuation OR funding OR "public offering")',
    "OPENAI": '"OpenAI" (IPO OR valuation OR funding OR "public offering")',
    "ANDURIL": '"Anduril" (valuation OR funding OR contract OR IPO)',
    "NEURALINK": '"Neuralink" (valuation OR funding OR trial OR IPO)',
    "SPACEX": '"SpaceX" (valuation OR funding OR IPO OR Starlink)',
    "POLYMARKET": '"Polymarket" (valuation OR funding OR regulation)',
    "FIGUREAI": '"Figure AI" (valuation OR funding OR robot)',
    "XAI": '"xAI" (valuation OR funding OR Grok)',
}
NAME = {"ANTHROPIC": "Anthropic", "OPENAI": "OpenAI", "ANDURIL": "Anduril",
        "NEURALINK": "Neuralink", "SPACEX": "SpaceX", "POLYMARKET": "Polymarket",
        "FIGUREAI": "Figure", "XAI": "xAI"}


def fetch(query):
    """One query, retried while the service says it is busy.

    It answers a refusal as plain text rather than as an error status, so the
    body decides whether this worked, not the status code.
    """
    url = (f"{GDELT}?query={urllib.parse.quote(query)}&mode=artlist"
           f"&maxrecords=60&format=json&sort=datedesc&timespan=7d")
    for attempt in range(TRIES):
        try:
            req = urllib.request.Request(url, headers=UA)
            with urllib.request.urlopen(req, timeout=45) as r:
                body = r.read().decode("utf-8", "replace")
            if body.lstrip().startswith("{"):
                return json.loads(body).get("articles") or []
        except (urllib.error.URLError, TimeoutError, json.JSONDecodeError):
            pass
        time.sleep(GAP * (attempt + 1))
    return None


def parse_stamp(s):
    """Their stamps look like 20260922T233000Z."""
    try:
        return int(time.mktime(time.strptime(s, "%Y%m%dT%H%M%SZ")) - time.timezone)
    except (ValueError, TypeError):
        return None


def tidy(articles, name):
    """One article per headline, and only where the headline names the company.

    A piece that mentions the company in passing is not news about it, and a
    rail full of general artificial intelligence stories would be noise beside
    a price.
    """
    named = re.compile(re.escape(name), re.I)
    seen, out = set(), []
    for a in articles or []:
        title = (a.get("title") or "").strip()
        key = title[:60].lower()
        if not title or not a.get("url") or key in seen or not named.search(title):
            continue
        seen.add(key)
        out.append({
            "title": re.sub(r"\s+", " ", title),
            "url": a["url"],
            "domain": a.get("domain"),
            "image": a.get("socialimage") or None,
            "at": parse_stamp(a.get("seendate")),
        })
    out.sort(key=lambda x: x["at"] or 0, reverse=True)
    return out


def build():
    out, refused = {}, []
    for i, (symbol, query) in enumerate(QUERY.items()):
        if i:
            time.sleep(GAP)
        articles = fetch(query)
        if articles is None:
            refused.append(symbol)
            continue
        out[symbol] = tidy(articles, NAME[symbol])[:12]
        print(f"  {symbol:10} {len(out[symbol]):>2} headlines", flush=True)

    # Anything refused is asked again, once, after a longer wait. Being turned
    # away is not the same as there being no news, and the page would otherwise
    # say so for a whole hour.
    if refused:
        print(f"  waiting, then asking again for {', '.join(refused)}", flush=True)
        time.sleep(GAP * 4)
        still = []
        for symbol in refused:
            articles = fetch(QUERY[symbol])
            if articles is None:
                still.append(symbol)
                continue
            out[symbol] = tidy(articles, NAME[symbol])[:12]
            print(f"  {symbol:10} {len(out[symbol]):>2} headlines on the second ask", flush=True)
            time.sleep(GAP)
        refused = still

    # A run that got nothing must not replace a good file with an empty one.
    path = os.path.join(HERE, "site", "news.json")
    if not out:
        print("  the news service refused every query, keeping what is there")
        return None
    if refused and os.path.exists(path):
        old = json.load(open(path)).get("by_symbol", {})
        for symbol in refused:
            if old.get(symbol):
                out[symbol] = old[symbol]
                print(f"  {symbol:10} kept {len(out[symbol])} from the last run")

    data = {"read_at": int(time.time()), "by_symbol": out,
            "source": "GDELT Project, the global news index",
            "refused": refused}
    with open(path, "w") as f:
        json.dump(data, f, indent=1)
    print(f"news written, {sum(len(v) for v in out.values())} headlines across "
          f"{len(out)} companies")
    return data


if __name__ == "__main__":
    build()

File: app/site/news.js
/* Headlines about the companies these markets price.

   A number that moves should be explainable. When Anthropic's bracket
   reprices, the reason is usually in the news that day, so the terminal puts
   the two beside each other and lets the reader judge.

   Articles come from GDELT, which indexes news worldwide and answers browser
   requests directly. It asks for no more than one query every five seconds, so
   queries are queued and spaced, and answers are kept for a while rather than
   asked for again.

   Nothing here is summarised or rewritten. A headline, its publisher and its
   time, linking out to the original.
*/
const GDELT = "https://api.gdeltproject.org/api/v2/doc/doc";
const GAP_MS = 6000;          // their limit is one every five seconds
const TRIES = 3;              // a busy answer is retried, not surfaced as an error
const KEEP_MS = 10 * 60_000;  // an answer stays good for ten minutes

/* Narrow the query to the company as a business story, otherwise a search for
   Anthropic returns every article that mentions artificial intelligence. */
const QUERY = {
  ANTHROPIC: '"Anthropic" (IPO OR valuation OR funding OR "public offering")',
  OPENAI: '"OpenAI" (IPO OR valuation OR funding OR "public offering")',
  ANDURIL: '"Anduril" (valuation OR funding OR contract OR IPO)',
  NEURALINK: '"Neuralink" (valuation OR funding OR trial OR IPO)',
  SPACEX: '"SpaceX" (valuation OR funding OR IPO OR Starlink)',
  POLYMARKET: '"Polymarket" (valuation OR funding OR regulation)',
  FIGUREAI: '"Figure AI" (valuation OR funding OR robot)',
  XAI: '"xAI" (valuation OR funding OR Grok)',
};

/* The name a headline has to carry for the story to be about this company. */
const NAME = {
  ANTHROPIC: "Anthropic", OPENAI: "OpenAI", ANDURIL: "Anduril",
  NEURALINK: "Neuralink", SPACEX: "SpaceX", POLYMARKET: "Polymarket",
  FIGUREAI: "Figure", XAI: "xAI",
};

const cache = new Map();      // symbol -> {at, articles}
let queue = Promise.resolve(), lastCall = 0;

/* One at a time, never faster than their limit allows. */
function spaced(fn){
  queue = queue.then(async () => {
    const wait = Math.max(0, GAP_MS - (Date.now() - lastCall));
    if(wait) await new Promise(r => setTimeout(r, wait));
    lastCall = Date.now();
    return fn();
  }).catch(() => null);
  return queue;
}

function tidy(articles, name){
  const seen = new Set();
  const named = new RegExp(name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), "i");
  return (articles || []).filter(a => {
    /* GDELT returns the same story from every site that syndicated it, so one
       per headline. And an article that merely mentions the company somewhere
       is not news about it: the headline has to name it. */
    const key = (a.title || "").slice(0, 60).toLowerCase();
    if(!a.title || !a.url || seen.has(key) || !named.test(a.title)) return false;
    seen.add(key);
    return true;
  }).map(a => ({
    title: a.title.replace(/\s+/g, " ").trim(),
    url: a.url,
    domain: a.domain,
    image: a.socialimage || null,
    at: parse(a.seendate),
  })).sort((x, y) => y.at - x.at);
}

/* GDELT stamps articles as 20260922T233000Z. */
function parse(s){
  if(!s || s.length < 15) return Date.now();
  const iso = `${s.slice(0,4)}-${s.slice(4,6)}-${s.slice(6,8)}T` +
              `${s.slice(9,11)}:${s.slice(11,13)}:${s.slice(13,15)}Z`;
  const t = Date.parse(iso);
  return isNaN(t) ? Date.now() : t;
}

export async function headlines(symbol, max = 8){
  const q = QUERY[symbol];
  if(!q) return [];
  const hit = cache.get(symbol);
  if(hit && Date.now() - hit.at < KEEP_MS) return hit.articles.slice(0, max);

  const name = NAME[symbol] || symbol;
  const articles = await spaced(async () => {
    const url = `${GDELT}?query=${encodeURIComponent(q)}&mode=artlist` +
                `&maxrecords=60&format=json&sort=datedesc&timespan=7d`;
    /* Their limit is per address, so a visitor who opens several companies at
       once can be turned away. That is a wait, not a failure. */
    for(let attempt = 0; attempt < TRIES; attempt++){
      const r = await fetch(url, {cache: "no-store"});
      const text = r.ok ? await r.text() : "";
      if(text.trim().startsWith("{")) return tidy(JSON.parse(text).articles, name);
      await new Promise(res => setTimeout(res, GAP_MS * (attempt + 1)));
    }
    throw new Error("news service stayed busy");
  });

  if(!articles) return hit ? hit.articles.slice(0, max) : [];
  cache.set(symbol, {at: Date.now(), articles});
  return articles.slice(0, max);
}
